### PR TITLE
CI: Trigger workflows based on changed files

### PR DESCRIPTION
**Pull request by seisman**
Some workflow runs are unnecessary and waste CI resources. For example, PR #2950 contains no codes or docs changes, but when it was merged into the main branch, workflows "Tests" and "Docs" were still triggered (see https://github.com/GenericMappingTools/pygmt/actions?query=branch%3Amain).

This PR updates the workflow trigger conditions based on changes files, to save CI resources.

- [x] benchmarks.yml
- [x] cache_data.yaml
- [x] check-links.yml
- [x] ci_docs.yml
- [x] ci_doctests.yaml
- [x] ci_tests_dev.yaml
- [x] ci_tests_legacy.yaml
- [x] ci_tests.yaml
- [x] dvc-diff.yml
- [x] format-command.yml
- [x] publish-to-pypi.yml
- [x] release-baseline-images.yml
- [x] release-drafter.yml
- [x] slash-command-dispatch.yml
- [x] style_checks.yaml
- [x] type_checks.yml

### CODE REVIEW

File: .github/workflows/publish-to-pypi.yml
Comment (seisman) [inline .github/workflows/publish-to-pypi.yml:20]: Only publish a test release to TestPyPI if there are code changes.

File: .github/workflows/benchmarks.yml
Comment (weiji14) [reply .github/workflows/benchmarks.yml:12]: Yes, #2945 was a good example of this where we only modified the test files, and that caused a change in performance. But this is a bit of a rare case though. Maybe we only need to run the benchmark on merge into main, but not in the Pull Request itself? I.e. delete L14 but keep L19?

File: .github/workflows/ci_docs.yml
Comment (weiji14) [inline .github/workflows/ci_docs.yml:27]: Maybe add `ci/requirements/docs.yml` too
```suggestion
      - '.github/workflows/ci_docs.yml'
      - 'ci/requirements/docs.yml'
```

File: .github/workflows/ci_docs.yml
Comment (weiji14) [inline .github/workflows/ci_docs.yml:36]:
```suggestion
      - '.github/workflows/ci_docs.yml'
      - 'ci/requirements/docs.yml'
```

File: .github/workflows/publish-to-pypi.yml
Comment (weiji14) [inline .github/workflows/publish-to-pypi.yml:26]: Reminder to change this to 'README.md' once https://github.com/GenericMappingTools/pygmt/issues/2928 is done.

File: .github/workflows/benchmarks.yml
Comment (seisman) [reply .github/workflows/benchmarks.yml:12]: > Maybe we only need to run the benchmark on merge into main, but not in the Pull Request itself?

Or just disable the workflow in PR and run it via slash command when necessary?

File: .github/workflows/ci_docs.yml
Comment (seisman) [reply .github/workflows/ci_docs.yml:27]: `ci/requirements/docs.yml` is only used by ReadTheDocs, so it's not related to this workflow.

File: .github/workflows/benchmarks.yml
Comment (weiji14) [reply .github/workflows/benchmarks.yml:12]: > Or just disable the workflow in PR (i.e. remove lines 16-20) and run it via slash command when necessary?

Yeah, let's go with that. Just keep a commented line like `# pull_request:` too, so we can manually trigger the benchmark workflow if needed. The slash command can be done in a separate PR.

File: .github/workflows/ci_docs.yml
Comment (weiji14) [reply .github/workflows/ci_docs.yml:27]: Oh right, ignore this then.

File: .github/workflows/benchmarks.yml
Comment (seisman) [reply .github/workflows/benchmarks.yml:12]: Done in b1e5fa81bf4e7d097ebcd2f76311128f62ec60ae.

File: .github/workflows/publish-to-pypi.yml
Comment (seisman) [inline .github/workflows/publish-to-pypi.yml:22]: Need to exclude `pygmt/tests/**` if we make the decision in #2956.

Comment (weiji14) [reply .github/workflows/publish-to-pypi.yml:22]: Add this after merging #2957

```suggestion
      - 'pygmt/**'
      - '!pygmt/tests/**'
```

File: .github/workflows/benchmarks.yml
Comment (weiji14) [inline .github/workflows/benchmarks.yml:15]: Add our standard hint about uncommenting the pull_request line
```suggestion
  # Uncomment the 'pull_request' line below to trigger the workflow in PR
  # pull_request:
```

File: .github/workflows/ci_tests_legacy.yaml
Comment (weiji14) [inline .github/workflows/ci_tests_legacy.yaml:14]:
```suggestion
  # Uncomment the 'pull_request' line below to trigger the workflow in PR
  # pull_request:
```

**Review by weiji14 — APPROVED** [review]
Just a couple more suggestions, and should be ok to merge after handling https://github.com/GenericMappingTools/pygmt/pull/2952/files#r1443654304.

**Pull request merged**
